File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static int[][] visited;
    static int[] dx = {1,-1,0,0};
    static int[] dy = {0,0,1,-1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        visited = new int[maps.length][maps[0].length];
        bfs(maps, visited);
        answer = visited[maps.length-1][maps[0].length-1];
        if(answer == 0){
            answer = -1;
        }
        return answer;
    }

    static void bfs(int[][] maps, int[][] visited){
        Queue<int[]> q = new LinkedList<>();
        visited[0][0] = 1;
        q.offer(new int[]{0,0});
        while(!q.isEmpty()){
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if(nx >= 0 && nx < maps.length && ny >= 0 && ny < maps[0].length){
                    if(visited[nx][ny] == 0 && maps[nx][ny] != 0){
                        visited[nx][ny] = visited[x][y] + 1;
                        q.offer(new int[]{nx,ny});
                    }
                }
            }
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(start: tuple, map: list):
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    r, c = len(map), len(map[0])
    q = deque([start])

    while q:
        x,y = q.popleft()

        # 벽 = 0
        # 길 = 1
        # 길에 누적하여 방문 추적
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < r and 0 <= ny < c and map[nx][ny] == 1:
                map[nx][ny] = map[x][y] + 1
                q.append((nx, ny))
                if nx == r-1 and ny == c-1:
                    return map[nx][ny]
    return -1
def solution(maps):  
    return bfs((0,0), maps)